src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs

```diff
@@ -447,14 +447,30 @@ internal TexlBinding ApplyBindingInternal()
                     bool notCoerceToType = false;
                     if (_allowCoerceToType)
                     {
-                        if (this._expectedReturnType != FormulaType.String)
+                        if (this._expectedReturnType == FormulaType.String)
                         {
-                            throw new NotImplementedException();
+                            if (!StringValue.AllowedListConvertToString.Contains(this.ReturnType))
+                            {
+                                notCoerceToType = true;
+                            }
                         }
-
-                        if (!StringValue.AllowedListConvertToString.Contains(this.ReturnType))
+                        else if (this._expectedReturnType == FormulaType.Number)
                         {
-                            notCoerceToType = true;
+                            if (!NumberValue.AllowedListConvertToNumber.Contains(this.ReturnType))
+                            {
+                                notCoerceToType = true;
+                            }
+                        }
+                        else if (this._expectedReturnType == FormulaType.Decimal)
+                        {
+                            if (!DecimalValue.AllowedListConvertToDecimal.Contains(this.ReturnType))
+                            {
+                                notCoerceToType = true;
+                            }
+                        }
+                        else
+                        {
+                            throw new NotImplementedException();
                         }

```

Comment:
I would write switch case for readability, something like

switch (this._expectedReturnType)
{
    case FormulaType.String:
        notCoerceToType = !StringValue.AllowedListConvertToString.Contains(this.ReturnType);
        break;
    case FormulaType.Number:
        notCoerceToType = !NumberValue.AllowedListConvertToNumber.Contains(this.ReturnType);
        break;
    case FormulaType.Decimal:
        notCoerceToType = !DecimalValue.AllowedListConvertToDecimal.Contains(this.ReturnType);
        break;
    default:
        throw new NotImplementedException();
}

Reply:
Updated this. thanks